server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageAnalyzer.java

```diff
@@ -0,0 +1,665 @@
+            for (Map.Entry<Integer, Long> field : visitor.fields.entrySet()) {
+                final String fieldName = fieldInfos.fieldInfo(field.getKey()).name;
+                final long fieldSize = (long) Math.ceil(field.getValue() * ratio);
+                stats.addStoredField(fieldName, fieldSize);
+            }
+        }
+    }
+
+    private static class TrackingSizeStoredFieldVisitor extends StoredFieldVisitor {
+        private final Map<Integer, Long> fields = new HashMap<>();
+
+        private void trackField(FieldInfo fieldInfo, int fieldLength) {
+            final int totalBytes = fieldLength + Long.BYTES; // a Long for bitsAndInfo
+            fields.compute(fieldInfo.number, (k, v) -> v == null ? totalBytes : v + totalBytes);
+        }
+
+        @Override
+        public void binaryField(FieldInfo fieldInfo, byte[] value) throws IOException {
+            trackField(fieldInfo, Integer.BYTES + value.length);
+        }
+
+        @Override
+        public void stringField(FieldInfo fieldInfo, byte[] value) throws IOException {
+            trackField(fieldInfo, Integer.BYTES + value.length);
+        }
+
+        @Override
+        public void intField(FieldInfo fieldInfo, int value) throws IOException {
+            trackField(fieldInfo, Integer.BYTES);
+        }
+
+        @Override
+        public void longField(FieldInfo fieldInfo, long value) throws IOException {
+            trackField(fieldInfo, Long.BYTES);
+        }
+
+        @Override
+        public void floatField(FieldInfo fieldInfo, float value) throws IOException {
+            trackField(fieldInfo, Float.BYTES);
+        }
+
+        @Override
+        public void doubleField(FieldInfo fieldInfo, double value) throws IOException {
+            trackField(fieldInfo, Double.BYTES);
+        }
+
+        @Override
+        public Status needsField(FieldInfo fieldInfo) throws IOException {
+            return Status.YES;
+        }
+    }
+
+    void analyzeDocValues(SegmentReader reader, IndexDiskUsageStats stats) throws IOException {
+        // TODO: We can extract these stats from Lucene80DocValuesProducer without iterating all docValues
+        // or track the new sliced IndexInputs.
+        DocValuesProducer docValuesReader = reader.getDocValuesReader();
+        if (docValuesReader == null) {
+            return;
+        }
+        docValuesReader = reader.getDocValuesReader().getMergeInstance();
+        for (FieldInfo field : reader.getFieldInfos()) {
+            final DocValuesType dvType = field.getDocValuesType();
+            if (dvType == DocValuesType.NONE) {
+                continue;
+            }
+            cancellationChecker.checkForCancellation();
+            directory.resetBytesRead();
+            switch (dvType) {
+                case NUMERIC:
+                    final NumericDocValues numeric = docValuesReader.getNumeric(field);
+                    while (numeric.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
```

Comment:
Just wondering: Could we cheat by jumping to `maxDoc * 0.9` using `advance` and then scan the rest with nextDoc if there's a match?

Reply:
Maybe even use binary search to figure out the last doc ID in the segment that has a value? (possibly starting with `maxDoc-1` for the common case when all docs have a value)

Reply:
Nice suggestion. I pushed https://github.com/elastic/elasticsearch/pull/74051/commits/ab2b0dd782a19139158882894b1f9cace955e2dd, which reduces the iterating time of docValues by more than 10 times.